internal/summary.go

```diff
@@ -0,0 +1,153 @@
+// This is necessary because we cannot declare a constant array in Go.
+func GetSeverityReportOrder() []querying.FindingSeverityType {
+	return []querying.FindingSeverityType{
+		querying.FindingSeverityCritical,
+		querying.FindingSeverityHigh,
+		querying.FindingSeverityModerate,
+		querying.FindingSeverityLow,
+		querying.FindingSeverityInfo,
+		querying.FindingSeverityUndefined,
+	}
+}
+
+type FindingSummary struct {
+	TotalCount       int
+	AffectedRepos    int
+	VulnsByEcosystem map[querying.FindingEcosystemType]int
+	VulnsBySeverity  map[querying.FindingSeverityType]int
+}
+
+type ProjectFindingSummary struct {
+	FindingSummary
+
+	Name string
+}
+
+// GetHighestCriticality looks for the severity level of the most critical
+// vulnerability in a project.
+func (r FindingSummary) GetHighestCriticality() querying.FindingSeverityType {
+	severities := GetSeverityReportOrder()
+	for _, sev := range severities {
+		count, exists := r.VulnsBySeverity[sev]
+		if exists && count > 0 {
+			return sev
+		}
+	}
+	return querying.FindingSeverityUndefined
+}
+
+func NewFindingSummary() FindingSummary {
+	return FindingSummary{
+		AffectedRepos:    0,
+		TotalCount:       0,
+		VulnsByEcosystem: map[querying.FindingEcosystemType]int{},
+		VulnsBySeverity:  NewSeverityMap(),
+	}
+}
+
+func NewProjectFindingSummary(name string) ProjectFindingSummary {
+	summary := NewFindingSummary()
+	return ProjectFindingSummary{Name: name, FindingSummary: summary}
+}
+
+func SummarizeFindings(projects *querying.ProjectCollection) (FindingSummary, []ProjectFindingSummary) {
+	affectedRepos, vulnCount := 0, 0
+	summary := NewFindingSummary()
+	projectReportCollection := []ProjectFindingSummary{}
+
+	for _, project := range projects.Projects {
+		projectReport := NewProjectFindingSummary(project.Name)
+		if numFindings := len(project.Findings); numFindings > 0 {
+			affectedRepos += 1
+			vulnCount += numFindings
+			projectReport.AffectedRepos = 1
+			projectReport.TotalCount = numFindings
+			// For each finding, add its ecosystem and severity tally to both
+			// the summary report and the project-specific report
+			for _, finding := range project.Findings {
+				summary.VulnsByEcosystem[finding.Ecosystem] += 1
+				summary.VulnsBySeverity[finding.Severity] += 1
+
+				projectReport.VulnsByEcosystem[finding.Ecosystem] += 1
+				projectReport.VulnsBySeverity[finding.Severity] += 1
+			}
+		}
+		projectReportCollection = append(projectReportCollection, projectReport)
+	}
+	summary.AffectedRepos = affectedRepos
+	summary.TotalCount = vulnCount
+
+	return summary, projectReportCollection
+}
+
+// TeamProjectCollection is a concrete type so that it can implement the sort
+// interface, for custom sorting.
+type TeamProjectCollection []*ProjectFindingSummary
+
+func (r TeamProjectCollection) Len() int {
+	return len(r)
+}
+
+func (r TeamProjectCollection) Swap(i, j int) {
+	r[i], r[j] = r[j], r[i]
```

Comment:
[q] this is safe without any placeholders, right?

Reply:
Placeholders for...?
This is basically a verbatim copy I found from every single example of implementing `sort.Interface` 😄 Even in the official docs! https://pkg.go.dev/sort#example-package

Reply:
Nope! You can do the same in Python too, matter of fact. It's a super nice shortcut. And an oft-used interview question as well. 😉